### PR TITLE
potential evapotraspiration BMI from alt-modular

### PR DESCRIPTION
**Pull request by jmframe**
[Short description explaining the high-level reason for the pull request]

## Additions

- All the stuff from the Alt-Modular repository.

## Removals

- Template material from repo setup

## Changes

- No changes from alt-modular

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,105 +1,80 @@
...
-3. Books, papers, talks, or other sources that have meaningful impact or influence on this project
+# Basic Model Interface (BMI) for potential evapotranspiration (PET) functions:
+* Aerodynamic method
+* Combination method
+* Energy balance method
+* Penman Monteith method
+* Priestley Taylor method
+
+# The configuration file (./configs/pet_config*.txt)
+Many options are available when running this PET module. One option is passed into the executable, and that is the speficif PET method option, the rest are passed in through the configuration file. These include forcing data (type and location), vegetation characteristics, site latitude/longitude/elevation, turbidity roughness and options about the forcings available vs calculated and assumed. Each instance of a PET model should have its own unique configuration file.
```
Comment (madMatchstick) [inline README.md:9]: typo: specific

File: README.md [hunk 2]
```diff
@@ -1,105 +1,80 @@
...
+* Priestley Taylor method
+
+# The configuration file (./configs/pet_config*.txt)
+Many options are available when running this PET module. One option is passed into the executable, and that is the speficif PET method option, the rest are passed in through the configuration file. These include forcing data (type and location), vegetation characteristics, site latitude/longitude/elevation, turbidity roughness and options about the forcings available vs calculated and assumed. Each instance of a PET model should have its own unique configuration file. 
+
+# Compiling this code
+The BMI functionality was developed as a standalone module in C. To compile this code the developer used these steps:
+1. `module load gnu/10.1.0`
+2. `gcc -lm ./src/pet.c ./src/bmi_pet.c -o run_bmi`
+This should generate an executable called **run_bmi**. To run this executable you must pass the path to corresponding to the corresponding configuration file, which includes the PET method you would like to run. Unit tests for those methods, and corresponding are provided, and can be run using:
```
Comment (madMatchstick) [inline README.md:15]: repeated phrase: path to ~~corresponding to~~ the corresponding configuration file

File: README.md [hunk 3]
```diff
@@ -1,105 +1,80 @@
...
+1. Energy balance method: `./run_bmi pet_config_unit_test1.txt`
+2. Aerodynamic method: `./run_bmi pet_config_unit_test2.txt`
+3. Combination method: `./run_bmi pet_config_unit_test3.txt`
+4. Priestley Taylor method: `./run_bmi pet_config_unit_test4.txt`
+5. Penman Monteith method: `./run_bmi pet_config_unit_test5.txt`
+Included in this repository is an environment file (env_cheyenne.sh), and a "make and run" file (make_and_run.sh), which will compile the code and run the five PET methods. If you are on the Cheyenne computer, or if you can modify these files to your machine, you can simply follow these two steps to run this code:
+1. `source env_cheyenne.sh`
+2. `./make_and_run.sh`
+
+# This rough code outline shows approximately what is happening.
```
Comment (madMatchstick) [inline README.md:25]: suggestion: Basic Outline of Workflow

File: README.md [hunk 4]
```diff
@@ -1,105 +1,80 @@
...
+Included in this repository is an environment file (env_cheyenne.sh), and a "make and run" file (make_and_run.sh), which will compile the code and run the five PET methods. If you are on the Cheyenne computer, or if you can modify these files to your machine, you can simply follow these two steps to run this code:
+1. `source env_cheyenne.sh`
+2. `./make_and_run.sh`
+
+# This rough code outline shows approximately what is happening. 
+The `pet_bmi.c` file runs BMI functions that initialize, update and finalize an instance of a PET model. It also includes descriptive functions to interpret specifics of the model, such as variable names, units, time/timestep, etc. And it also allows a user (or framework) to get and set values in this model. The `bmi_pet.c` code interacts with the `pet.c` code, which sets up the model based on the PET method chosen. For instance, the aerodynamic method does not calculate the net radiation before calling the PET subroutine, but the other PET methods do. This pet.c file then calls one of the five PET methods available at this time. When the method is called they return a value for PET in m/s, and that is set directly to the BMI model structure.
```
Comment (madMatchstick) [inline README.md:26]: syntax: When ~~the~~ a method is called ~~they return~~ it returns a value for PET

File: README.md [hunk 5]
```diff
@@ -1,105 +1,80 @@
...
+![code_flow](./figs/bmi_pet.png)
+
+# Notes from author of the PET functions
+evapotranspiration (ET) module,  
+Version 1.0 by Fred L. Ogden, NOAA-NWS-OWP, May, 2020.  
+includes five different methods to clculate ET, from Chow, Maidment & Mays Textbook, and UNFAO Penman-Monteith:
```
Comment (madMatchstick) [inline README.md:32]: typo: calculate

File: README.md [hunk 6]
```diff
@@ -1,105 +1,80 @@
...
+4. Priestley-Taylor method, which assumes the ratio between 1 & 2, and only calculates 1.
+5. Penman-Monteith method, which requires a value of canopy resistance term, and does not rely on 1 or 2.
+This subroutine requires a considerable amount of meteorological data as input.
+   * a) temperature and (relative-humidity or specific humidity) and the heights at which they are measured.
+   * b) near surface wind speed measurement and the height at which it was measured.
+   * c) the ambient atmospheric temperature lapse rate
+   * d) the fraction of the sky covered by clouds
+   * e) (optional) the height above ground to the cloud base. If not provided, then assumed.
+   * f) the day of the year (1-366) and time of day (UTC only!)
+   * g) the skin temperature of the earth's surface, TODO: should come from another module to calc. soil or veg. temp.
```
Comment (madMatchstick) [inline README.md:45]: suggestion: remove TODO from readme but add as issue-enhancement to repo?

File: README.md [hunk 7]
```diff
@@ -1,105 +1,80 @@
...
+   * h) the zero-plane roughness height of the atmospheric boundary layer assuming log-law behavior (from land cover)
+   * i) the average root zone soil temperature, or near-surface water temperature in the case of lake evaporation.
+   * j) the incoming solar (shortwave) radiation.  If not provided it is computed from d,e,f, using an updated method similar to the one presented in Bras, R.L. Hydrology.  Requires value of the Linke atmospheric turbidity factor, which varies from 2 for clear mountain air to 5 for smoggy air.  According to Hove & Manyumbu 2012, who calculated values over Zimbabwe that varied from 2.14 to 3.71.  Other values exist in the literature. TODO: This turbidity factor could be calculated from satellite obs. or maybe NOAA already does this? All radiation calculations needed for 1, 3, 4, and 5 require net radiation calculations at the land surface. The net radiation is calculated using a, c, d, e, f, g, j, plus the Linke turbidity factor, which can be estimate from satellite observations.
```
Comment (madMatchstick) [inline README.md:48]: suggestion: remove TODO from readme but add as issue-enhancement to repo?

Comment (madMatchstick) [inline README.md:48]: typo: estimated

File: README.md [hunk 8]
```diff
@@ -1,105 +1,80 @@
...
+
+Roughness definitions according to the European Wind Atlas.  
+According to the UN FAO Penman-Monteith example here: http://www.fao.org/3/X0490E/x0490e06.htm#aerodynamic%20resistance%20(ra)  
+The zero plane roughness length,"d" can be approximated as 2/3 of the vegetation height (H): d=2/3*H.  
+The momentum roughness height "zom" can be estimated as 0.123*H.  
+The heat transfer roughness height "zoh" can be approximated as 0.1 * zom.  
+
+# A note on code adaptation for BMI
+This code was minimally changed from the author's original version. These minor changes were made by NGen NWM formulation team:
+* Much of this C code was moved to `*.h` files, with the intention of being more easily integrated into the NGen Framework. It turned out that this step was not stricktly neccessary, and that this standalone module could use the standard `*.c` files. See known issues below for a discussion on turning these back to `*.c` files.
```
Comment (madMatchstick) [inline README.md:73]: typos: strictly necessary

File: README.md [hunk 9]
```diff
@@ -1,105 +1,80 @@
...
+* At one point the C code was slightly modified to compile as C++ code. This was intended for easier integration with the NGen Framework. This was reversed when developed as a standalone module.
+* Functions that are neccessary for PET calculations, but are not part of an individual method (e.g., `calculate_solar_radiation`), were moved to this file: `./include/pet_tools.h`
```
Comment (madMatchstick) [inline README.md:75]: typo: necessary

File: README.md [hunk 10]
```diff
@@ -1,105 +1,80 @@
...
+| 4         | 1.6        | Very large cities with tall buildings and skyscrapers|  
+
+Roughness definitions according to the European Wind Atlas.  
+According to the UN FAO Penman-Monteith example here: http://www.fao.org/3/X0490E/x0490e06.htm#aerodynamic%20resistance%20(ra)  
+The zero plane roughness length,"d" can be approximated as 2/3 of the vegetation height (H): d=2/3*H.  
+The momentum roughness height "zom" can be estimated as 0.123*H.  
+The heat transfer roughness height "zoh" can be approximated as 0.1 * zom.  
+
+# A note on code adaptation for BMI
+This code was minimally changed from the author's original version. These minor changes were made by NGen NWM formulation team:
```
Comment (madMatchstick) [inline README.md:72]: suggestion: These minor changes were made by ~~NGen~~ Nextgen National Water Model (NWM) ~~f~~Formulation team

File: README.md (hunk 8 above)
Comment (madMatchstick) [inline README.md:73]: suggestion: Nextgen Framework

File: README.md [hunk 11]
```diff
@@ -1,105 +1,80 @@
...
+| 1         | 0.03       | Open agricultural area without fences and hedgerows and very scattered buildings. Only softly rounded hills |  
+| 1.5       | 0.055      | Agricultural land with some houses and 8 metre tall sheltering hedgerows with a distance of approximately 1250 metres |  
+| 2         | 0.1        | Agricultural land with some houses and 8 metre tall sheltering hedgerows with a distance of approximately 500 metres |  
+| 2.5       | 0.2        | Agricultural land with many houses, shrubs and plants, or 8 metre tall sheltering hedgerows with a distance of approximately 250 metres |  
+| 3         | 0.4        | Villages, small towns, agricultural land with many or tall sheltering hedgerows, forests and very rough and uneven terrain |  
+| 3.5       | 0.8        | Larger cities with tall buildings|  
+| 4         | 1.6        | Very large cities with tall buildings and skyscrapers|  
+
+Roughness definitions according to the European Wind Atlas.  
+According to the UN FAO Penman-Monteith example here: http://www.fao.org/3/X0490E/x0490e06.htm#aerodynamic%20resistance%20(ra)
```
Comment (madMatchstick) [inline README.md:66]: suggestion: add url as link [here].(http://www.fao.org/3/X0490E/x0490e06.htm#aerodynamic%20resistance%20(ra))

File: README.md [hunk 12]
```diff
@@ -1,105 +1,80 @@
...
+* Penman Monteith method
+* Priestley Taylor method
+
+# The configuration file (./configs/pet_config*.txt)
+Many options are available when running this PET module. One option is passed into the executable, and that is the speficif PET method option, the rest are passed in through the configuration file. These include forcing data (type and location), vegetation characteristics, site latitude/longitude/elevation, turbidity roughness and options about the forcings available vs calculated and assumed. Each instance of a PET model should have its own unique configuration file. 
+
+# Compiling this code
+The BMI functionality was developed as a standalone module in C. To compile this code the developer used these steps:
+1. `module load gnu/10.1.0`
+2. `gcc -lm ./src/pet.c ./src/bmi_pet.c -o run_bmi`
```
Comment (madMatchstick) [inline README.md:14]: missing`./src/main.c` ?

File: README.md [hunk 13]
```diff
@@ -1,105 +1,80 @@
...
+This should generate an executable called **run_bmi**. To run this executable you must pass the path to corresponding to the corresponding configuration file, which includes the PET method you would like to run. Unit tests for those methods, and corresponding are provided, and can be run using:
+1. Energy balance method: `./run_bmi pet_config_unit_test1.txt`
+2. Aerodynamic method: `./run_bmi pet_config_unit_test2.txt`
+3. Combination method: `./run_bmi pet_config_unit_test3.txt`
+4. Priestley Taylor method: `./run_bmi pet_config_unit_test4.txt`
+5. Penman Monteith method: `./run_bmi pet_config_unit_test5.txt`
+Included in this repository is an environment file (env_cheyenne.sh), and a "make and run" file (make_and_run.sh), which will compile the code and run the five PET methods. If you are on the Cheyenne computer, or if you can modify these files to your machine, you can simply follow these two steps to run this code:
+1. `source env_cheyenne.sh`
+2. `./make_and_run.sh`
```
Comment (madMatchstick) [inline README.md:23]: this file does not exist

File: make_and_run_pass_forcings.sh [hunk 14]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (madMatchstick) [inline make_and_run_pass_forcings.sh:1]: this runs okay

File: make_and_run_read_forcings.sh [hunk 15]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+gcc -lm ./src/main_read_forcing.c ./src/pet.c ./src/bmi_pet.c -o run_bmi_forcings_read
```
Comment (madMatchstick) [inline make_and_run_read_forcings.sh:2]: this runs okay

File: src/main.c [hunk 16]
```diff
@@ -0,0 +1,118 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "../include/bmi.h"
+#include "../include/et.h"
+#include "../include/bmi_et.h"
+
+#include "../../AORC/include/aorc.h"
```
Comment (madMatchstick) [inline src/main.c:8]: these two AORC paths need to be updated

File: src/main.c [hunk 17]
```diff
@@ -0,0 +1,118 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "../include/bmi.h"
+#include "../include/et.h"
```
Comment (madMatchstick) [inline src/main.c:5]: update include paths to `pet.h`

File: src/main.c [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+  if(argc<=1){
+    printf("make sure to include a path to the ET config file\n");
+    exit(1);
+  }
+
+  /************************************************************************
+      allocating memory to store the entire BMI structure for ET
+  ************************************************************************/
+  printf("allocating memory to store entire BMI ET structure\n");
+  Bmi *et_bmi_model = (Bmi *) malloc(sizeof(Bmi));
```
Comment (madMatchstick) [inline src/main.c:31]: update all refs from `et` to `pet`?

**Review by madMatchstick — CHANGES_REQUESTED** [review]
Having issues compiling code.  I believe the `main.c` needs attention?

File: README.md (hunk 7 above)
Comment (jmframe) [reply README.md:48]: Typo removed since moving the TODO to an issue. Thanks.

File: README.md (hunk 8 above)
Comment (jmframe) [reply README.md:73]: I kind of liked that spelling of 'stricktly'

File: README.md (hunk 10 above)
Comment (jmframe) [reply README.md:72]: Replaced all NGens with Nextgens.

File: README.md (hunk 8 above)
Comment (jmframe) [reply README.md:73]: Replaced all NGens with Nextgens.

File: src/main.c (hunk 16 above)
Comment (jmframe) [reply src/main.c:8]: Updated. But that main.c file might not be needed...

File: src/main.c (hunk 17 above)
Comment (jmframe) [reply src/main.c:5]: Updated. But that main.c file might not be needed... Might delete...

File: src/main.c (hunk 18 above)
Comment (jmframe) [reply src/main.c:31]: Yeah, will just delete that file.

File: README.md [hunk 19]
```diff
@@ -1,105 +1,80 @@
...
+The BMI functionality was developed as a standalone module in C. To compile this code the developer used these steps:
+1. `module load gnu/10.1.0`
+2. `gcc -lm ./src/main_read_forcing.c ./src/pet.c ./src/bmi_pet.c -o run_bmi`
+This should generate an executable called **run_bmi**. To run this executable you must pass the path to the corresponding configuration file, which includes the PET method you would like to run. Unit tests for those methods, and corresponding are provided, and can be run using:
+1. Energy balance method: `./run_bmi pet_config_unit_test1.txt`
+2. Aerodynamic method: `./run_bmi pet_config_unit_test2.txt`
+3. Combination method: `./run_bmi pet_config_unit_test3.txt`
+4. Priestley Taylor method: `./run_bmi pet_config_unit_test4.txt`
+5. Penman Monteith method: `./run_bmi pet_config_unit_test5.txt`
+Included in this repository is an environment file (env_cheyenne.sh), and a "make and run" file (make_and_run.sh), which will compile the code and run the five PET methods. If you are on the Cheyenne computer, or if you can modify these files to your machine, you can simply follow these two steps to run this code:
```
Comment (madMatchstick) [inline README.md:21]: add newline space between ln 20 & 21 to format [list](https://github.com/NOAA-OWP/evapotranspiration/blob/4f04e161893a0ec24b6b818e9c08399931110270/README.md#compiling-this-code) correctly.

File: README.md [hunk 20]
```diff
@@ -1,105 +1,80 @@
...
+1. `source env_cheyenne.sh`
+2. `./make_and_run_read_forcings.sh`
```
Comment (madMatchstick) [inline README.md:23]: What about `make_and_run_pass_forcings.sh`?  Best to briefly refer/comment on this script somewhere in README, unless I overlooked?

**Review by madMatchstick — APPROVED** [review]
Both build run scripts look good.  See 2 more minor comments in README.  Otherwise, merge is good to go :)

File: README.md (hunk 20 above)
Comment (jmframe) [reply README.md:23]: Added:
Included in this repository is an environment file (env_cheyenne.sh), and two "make and run" files. `make_and_run_read_forcings.sh` will compile the code and run the five PET methods. 
The other is `make_and_run_pass_forcings.sh`, which is an example of how the PET model will be run in a framework, where the forcing values are passed in through BMI functions (e.g., set_value()). An environment file exists, which should allow you to run compile and run the potential evaporation routines: `source env_cheyenne.sh`. Then all you need to do is run one or both of the two scrips:
1. `./make_and_run_read_forcings.sh`  
2. `./make_and_run_pass_forcings.sh`

**Pull request merged**